### PR TITLE
Add support for custom themes to use dark mode

### LINKED ISSUE #6588: Allow custom themes to have light and dark variants
## The request

0.114 has an "auto" mode for the default theme which lets it show a light or dark variant depending on your OS appearance.  Really nice and useful. Would be cool if support for this was extended to custom themes as well.

Many theme creators already offer different variants of their themes for light and dark mode. For example, here is a collection of 20 [JuanMTech themes](https://www.juanmtech.com/themes-in-home-assistant/). It's really ten themes with light and dark variants, though. 

So the idea here is rather than having "MyCoolCustomTheme - Light" and "MyCoolCustomTheme - Dark", there would instead be a single "MyCoolCustomTheme" with light and dark variants that get intelligently switched based on OS appearance, much like the default theme. Idea of how the YAML could look:

```
MyCoolCustomTheme:
  dark_mode:
    primary-background-color: "black"
    primary-text-color: "white"
  light_mode:
    primary-background-color: "white"
    primary-text-color: "black"
 ```

Some benefits:
* This would make custom themes behave more intelligently
* Makes it easier for users to find and install themes. Less searching, less copy and pasting, etc.
* Also makes it easier for users to update their themes when the author updates them (or they break due to HA update etc)
* Would decrease the "Theme" dropdown in the users profile page in half, making it easier for them to find/switch between their themes.

### PR DESCRIPTION
**Pull request by spacegaier**
## Breaking change

Not breaking, since legacy theme (= implicit light mode only) format is still supported.

## Proposed change

Backend PR: https://github.com/home-assistant/core/pull/46532

Adds the option for custom themes to be based upon the dark mode. A theme can now specify support for both light and dark mode and the user can select which one to use (same behavior as for the "default" theme mode selection). 

This PR also renames the local storage parameter "selectedTheme" to "selectedThemeSettings" (to underscore that it does not only contain a theme name, but all related user defined settings) including an automatic migration for the old key storage.

<s>This change also allows the user to specify default color values for light and dark mode for `primary_color` and `accent-color`. If provided, the same color pickers currently used by the "default" theme will be shown and users can adjust as desired.</s>
Reverted from MVP

11A99702-3F01-4618-AD48-B9E5AEC8EE96

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

Below the different test themes used in the GIF above to go through the different scenarios.

```yaml
themes:
    orange_dummy:
      primary-color: "#ff9800"
      primary-text-color: "#ff9800"
      secondary-color: "#ff9800" 
    green_dummy:
      modes:
        light:
          primary-color: "#00aa00"
          primary-text-color: "#00aa00"
          secondary-color: "#00aa00"
        dark:
          primary-color: "#44ff44"
          primary-text-color: "#44ff44"
          secondary-color: "#44ff44"     
          app-header-text-color: "#44ff44"     
    color_picker:
      primary-color: "#ff0000"
      accent-color: "#0000ff"
      modes:
        light:
          secondary-color: "#ffffff"
        dark:
          secondary-color: "#44ff44"     
          app-header-text-color: "#44ff44"     
    dark_only:
      modes:
        dark:  
          primary-color: "#ff00ff"
    light_only:
      modes:
        light:  
          primary-color: "#0000ff"
    blur:
      modes:
        light:
          dialog-backdrop-filter: blur(5px)
          iron-overlay-backdrop-opacity: 1
          iron-overlay-backdrop-background-color: rgba(0, 0, 0, 0.32)  
        dark:
          dialog-backdrop-filter: blur(5px)
          iron-overlay-backdrop-opacity: 1
          iron-overlay-backdrop-background-color: rgba(0, 0, 0, 0.32)  
```

## Additional information

- This PR fixes or closes issue: fixes https://github.com/home-assistant/frontend/issues/6588
- This PR is related to issue or discussion:
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/17960

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/dom/apply_themes_on_element.ts [hunk 1]
```diff
@@ -23,57 +23,104 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
-      themeRules["primary-color"] = themeOptions.primaryColor;
-      const rgbLigthPrimaryColor = lab2rgb(labBrighten(labPrimaryColor));
-      themeRules["light-primary-color"] = rgb2hex(rgbLigthPrimaryColor);
+      themeRules["primary-color"] = primaryColor;
+      const rgbLightPrimaryColor = lab2rgb(labBrighten(labPrimaryColor));
+      themeRules["light-primary-color"] = rgb2hex(rgbLightPrimaryColor);
       themeRules["dark-primary-color"] = lab2hex(labDarken(labPrimaryColor));
       themeRules["text-primary-color"] =
-        rgbContrast(rgbPrimaryColor, [33, 33, 33]) < 6 ? "#fff" : "#212121";
+        rgbContrast(rgbPrimaryColor, [33, 33, 33]) < 7 ? "#fff" : "#212121";
```
Comment (spacegaier) [inline src/common/dom/apply_themes_on_element.ts:91]: Previously we only executed this calculation if we had a user overridden primary color. That meant that if the user did not change it, we would stay with the default white on blue color for the header (since that is using `text-primary-color`). However those default color values do not follow the contrast value 6. To keep the look consistent with how it was before for the default theme I adjusted to 7. That however could mean that for users that manually specified a color for the default theme right at the cutoff point between contrast 6 and 7, their header color might change.

Alternative is to add a guard around  `text-primary-color` and related keys again to only execute it if the user overrides the colors in the picker. However, even only moving the default color by a single RGB value (smallest possible change), means that then this contrast logic takes effect and switches from white on blue to black on blue in the header, so we would only be hiding the fact that our own default colors do not conform with our defined desired contrast value.

File: src/components/map/ha-location-editor.ts [hunk 2]
```diff
@@ -109,7 +109,7 @@ class LocationEditor extends LitElement {
 
     if (changedProps.has("hass")) {
       const oldHass = changedProps.get("hass") as HomeAssistant | undefined;
-      if (!oldHass || oldHass.themes?.darkMode === this.hass.themes?.darkMode) {
```
Comment (bramkragten) [inline src/components/map/ha-location-editor.ts:112]: Why is this no longer needed? We still init `themes` as `null`?

File: src/util/ha-pref-storage.ts [hunk 3]
```diff
@@ -25,13 +31,17 @@ export function storeState(hass: HomeAssistant) {
...
 
-  STORED_STATE.forEach((key) => {
+  STORED_STATE.concat(STORED_STATE_DEPRECATED).forEach((key) => {
     if (key in STORAGE) {
       let value = JSON.parse(STORAGE[key]);
       // selectedTheme went from string to object on 20200718
       if (key === "selectedTheme" && typeof value === "string") {
         value = { theme: value };
       }
+      // selectedTheme was renamed to selectedThemeSettings on 20210207
```
Comment (bramkragten) [inline src/util/ha-pref-storage.ts:41]: Why did you rename this?

File: src/state/connection-mixin.ts [hunk 4]
```diff
@@ -34,6 +34,7 @@ export const connectionMixin = <T extends Constructor<HassBaseEl>>(
         states: null as any,
         config: null as any,
         themes: null as any,
+        selectedThemeSettings: null,
```
Comment (bramkragten) [inline src/state/connection-mixin.ts:37]: Shouldnt `selectedTheme` be removed if we are going to rename it?

Comment (spacegaier) [reply src/state/connection-mixin.ts:37]: `selectedTheme` is not this structure init, therefore cannot be removed here. Or am I missing your point?

File: src/components/map/ha-location-editor.ts (hunk 2 above)
Comment (spacegaier) [reply src/components/map/ha-location-editor.ts:112]: Should not have been needed before the PR either. Basically no usages of `hass.themes` do check via `?` (e.g. none of the cards do).

File: src/common/dom/apply_themes_on_element.ts (hunk 1 above)
Comment (spacegaier) [reply src/common/dom/apply_themes_on_element.ts]: Reverted since color picker support was removed and so this contrast issue is now less likely to be noticed (although I still think the old logic is flawed).

File: src/common/dom/apply_themes_on_element.ts [hunk 5]
```diff
@@ -23,57 +27,102 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+
+  if (themeSettings) {
+    // Determine the primary and accent colors. Fallbacks are the HA default blue and orange
+    // or the derived "darkStyles" values, depending on the light vs dark mode.
+    let primaryColor;
+    let accentColor;
+    if (selectedTheme === "default") {
+      // User selected colors
+      primaryColor = themeSettings.primaryColor || DEFAULT_PRIMARY_COLOR;
+      accentColor = themeSettings.accentColor || DEFAULT_ACCENT_COLOR;
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:51]:
```suggestion
      primaryColor = themeSettings.primaryColor;
      accentColor = themeSettings.accentColor;
```

File: src/common/dom/apply_themes_on_element.ts [hunk 6]
```diff
@@ -23,57 +27,102 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
-          "#121212",
-          8
-        );
-      }
+
+      themeRules["app-header-background-color"] = hexBlend(
+        primaryColor,
+        "#121212",
+        8
+      );
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:84]:
```suggestion
      if (primaryColor) {
      themeRules["app-header-background-color"] = hexBlend(
        primaryColor,
        "#121212",
        8
      );
      }
```

File: src/common/dom/apply_themes_on_element.ts [hunk 7]
```diff
@@ -23,57 +27,102 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+      accentColor = themeSettings.accentColor || DEFAULT_ACCENT_COLOR;
+    } else if (selectedTheme && themes.themes[selectedTheme]) {
+      // Try in that order:
+      // 1. Fixed values from theme styles
+      // 2. HA default colors
+      if (themeSettings.dark) {
+        primaryColor =
+          themes.themes[selectedTheme].styles?.dark!["primary-color"] ||
+          darkStyles["primary-color"] ||
+          DEFAULT_PRIMARY_COLOR;
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:60]:
```suggestion
        primaryColor =
          themes.themes[selectedTheme].styles?.dark!["primary-color"];
```

File: src/common/dom/apply_themes_on_element.ts [hunk 8]
```diff
@@ -23,57 +27,102 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+        accentColor =
+          themes.themes[selectedTheme].styles?.dark!["accent-color"] ||
+          darkStyles["accent-color"] ||
+          DEFAULT_ACCENT_COLOR;
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:64]:
```suggestion
        accentColor =
          themes.themes[selectedTheme].styles?.dark!["accent-color"];
```

File: src/common/dom/apply_themes_on_element.ts [hunk 9]
```diff
@@ -23,57 +27,102 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+      } else {
+        // eslint-disable-next-line no-lonely-if
+        primaryColor =
+          themes.themes[selectedTheme].styles?.light!["primary-color"] ||
+          DEFAULT_PRIMARY_COLOR;
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:69]:
```suggestion
        primaryColor =
          themes.themes[selectedTheme].styles?.light!["primary-color"];
```

File: src/common/dom/apply_themes_on_element.ts [hunk 10]
```diff
@@ -23,57 +27,102 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+        accentColor =
+          themes.themes[selectedTheme].styles?.light!["accent-color"] ||
+          DEFAULT_ACCENT_COLOR;
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:72]:
```suggestion
        accentColor =
          themes.themes[selectedTheme].styles?.light!["accent-color"];
```

File: src/common/dom/apply_themes_on_element.ts [hunk 11]
```diff
@@ -78,7 +113,17 @@ export const applyThemesOnElement = (
...
+    const { modes, ...baseThemeRules } = themes.themes[selectedTheme];
+    themeRules = { ...themeRules, ...baseThemeRules };
+
+    // If dark is requested, check if the theme actually supports dark mode. Otherwise we
+    // fall back to the light mode.
+    if (themeSettings?.dark && modes?.dark)
+      themeRules = { ...themeRules, ...modes?.dark };
+    else {
+      themeRules = { ...themeRules, ...modes?.light };
+    }
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:126]:
```suggestion
    if (modes) {
	      if (themeSettings?.dark) {
	        themeRules = { ...themeRules, ...modes.dark };
	      } else {
	        themeRules = { ...themeRules, ...modes.light };
	      }
    }
```

File: src/common/dom/apply_themes_on_element.ts [hunk 12]
```diff
@@ -23,50 +23,85 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+        accentColor =
+          themes.themes[selectedTheme].modes?.light?.["accent-color"];
+      }
+
+      // Mode-independent theme value as fallback
+      primaryColor =
+        primaryColor || themes.themes[selectedTheme]["primary-color"];
+      accentColor =
+        primaryColor || themes.themes[selectedTheme]["accent-color"];
+    }
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:70]:
```suggestion
  if (selectedTheme === "default" && themeSettings) {
    // Determine the primary and accent colors. Fallbacks are the HA default blue and orange
    // or the derived "darkStyles" values, depending on the light vs dark mode.
    const primaryColor = themeSettings.primaryColor;
    const accentColor = themeSettings.accentColor;
  }
```

File: src/common/dom/apply_themes_on_element.ts [hunk 13]
```diff
@@ -23,50 +23,85 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+
+    if (themeSettings.dark) {
       cacheKey = `${cacheKey}__dark`;
       themeRules = darkStyles;
-      if (themeOptions.primaryColor) {
+
+      if (primaryColor) {
         themeRules["app-header-background-color"] = hexBlend(
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:77]: Before doing this we should check if the theme doesn't contain a `app-header-background-color` value that will overrule this anyway

File: src/common/dom/apply_themes_on_element.ts [hunk 14]
```diff
@@ -23,62 +23,86 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
-      return;
+    if (selectedTheme === "default") {
+      // Determine the primary and accent colors from the current settings.
+      // Fallbacks are implicitly the HA default blue and orange or the
+      // derived "darkStyles" values, depending on the light vs dark mode.
+      const primaryColor = themeSettings.primaryColor;
+      const accentColor = themeSettings.accentColor;
+
+      if (themeSettings.dark) {
+        if (primaryColor && !("app-header-background-color" in themeRules)) {
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:53]:
```suggestion
      if (themeSettings.dark && primaryColor) {
```

File: src/common/dom/apply_themes_on_element.ts [hunk 15]
```diff
@@ -23,62 +23,86 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+          themeRules["app-header-background-color"] = hexBlend(
+            primaryColor,
+            "#121212",
+            8
+          );
+        }
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:59]:
```suggestion
```

File: src/common/dom/apply_themes_on_element.ts [hunk 16]
```diff
@@ -23,62 +23,86 @@ let PROCESSED_THEMES: Record<string, ProcessedTheme> = {};
...
+      }
+
+      // Nothing was changed
+      if (element._themes?.cacheKey === cacheKey) {
+        return;
+      }
     }
   }
 
   if (selectedTheme && themes.themes[selectedTheme]) {
```
Comment (bramkragten) [inline src/common/dom/apply_themes_on_element.ts:93]:
```suggestion
  else if (selectedTheme && themes.themes[selectedTheme]) {
```

Comment (bramkragten) [reply src/common/dom/apply_themes_on_element.ts:93]: We should only do this for custom themes, otherwise, it would overrule the calculations from above.

**Review by bramkragten — APPROVED** [review]
OK to merge when documentation is added
